File: DataStructure/double_ended_queue/double_ended_queue.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

typedef struct _node* nodePointer;
typedef struct _deque* dequePointer;

typedef struct _node {
	nodePointer prev;
	nodePointer next;
	int data;
}node;

typedef struct _deque {
	nodePointer rear;
	nodePointer front;
	int nodeCount;
}deque;

void queueInit(dequePointer q) {
	q->front = NULL;
	q->rear = NULL;
	q->nodeCount = 0;
}

bool isEmpty(dequePointer q) {
	if (q->rear == NULL && q->front == NULL)
		return true;
	else
		return false;
}

void rear_enQueue(dequePointer q, int data) {
	nodePointer newNode = (nodePointer)malloc(sizeof(node));
	newNode->data = data;
	newNode->prev = NULL;
	newNode->next = NULL;
	if (isEmpty(q)) {
		q->rear = newNode;
		q->front = newNode;
	}
	else {
		q->rear->prev = newNode;
		newNode->next = q->rear;
		q->rear = newNode;
	}
	q->nodeCount++;
}

void front_enQueue(dequePointer q, int data) {
	nodePointer newNode = (nodePointer)malloc(sizeof(node));
	newNode->data = data;
	newNode->prev = NULL;
	newNode->next = NULL;
	if (isEmpty(q)) {
		q->rear = newNode;
		q->front = newNode;
	}
	else {
		q->front->next = newNode;
		newNode->prev = q->front;
		q->front = newNode;
	}
	q->nodeCount++;
}

void rear_deQueue(dequePointer q) {
	nodePointer tmp = q->rear;
	if (!isEmpty(q)) {
		tmp->next->prev = NULL;
		q->rear = tmp->next;
		free(tmp);
	}
	q->nodeCount--;
}

void front_dequeue(dequePointer q) {
	nodePointer tmp = q->front;
	if(!isEmpty(q)){
		tmp->prev->next = NULL;
		q->front = tmp->prev;
		free(tmp);
	}
	q->nodeCount--;
}

void print_queue(dequePointer q) {
	nodePointer tmp = q->rear;
	for (int i = 0; i < q->nodeCount; i++) {
		printf("%3d  ", tmp->data);
		tmp = tmp->next;
	}
	printf("\n");
}

int main() {
	int arr[] = { 5, 10, 15, 20, 2, 4, 6, 8 };
	dequePointer q = (dequePointer)malloc(sizeof(deque));
	queueInit(q);

	for (int i = 0; i < sizeof(arr) / sizeof(int); i++)
		rear_enQueue(q, arr[i]);
	print_queue(q);

	rear_deQueue(q);
	print_queue(q);
	front_dequeue(q);
	print_queue(q);

	front_enQueue(q, 70);
	print_queue(q);

	return 0;
}

File: DataStructure/double_ended_queue/double_ended_queue.java
package double_ended_queue;

class node{
	node prev;
	node next;
	int data;
	
	public node(int data) {
		this.prev = null;
		this.next = null;
		this.data = data;
	}
}

class Deque{
	node rear;
	node front;
	int nodeCount;
	
	public Deque() {
		this.rear = null;
		this.front = null;
		this.nodeCount = 0;
	}
	
	boolean isEmpty() {
		if(this.rear == null && this.front == null)
			return true;
		else
			return false;
	}
	void rear_Enqueue(int data) {
		node newNode = new node(data);
		if(this.isEmpty()) {
			this.rear = newNode;
			this.front = newNode;
		}
		else {
			this.rear.prev = newNode;
			newNode.next = this.rear;
			this.rear = newNode;
		}
		this.nodeCount++;
	}
	void front_Enqueue(int data) {
		node newNode = new node(data);
		if(this.isEmpty()) {
			this.rear = newNode;
			this.front = newNode;
		}
		else {
			this.front.next = newNode;
			newNode.prev = this.front;
			this.front = newNode;
		}
		this.nodeCount++;
	}
	void rear_Dequeue() {
		node tmp = this.rear;
		if(!this.isEmpty()) {
			tmp.next.prev = null;
			this.rear = tmp.next;
			tmp = null;
		}
		this.nodeCount--;
	}
	void front_Dequeue() {
		node tmp = this.front;
		if(!this.isEmpty()) {
			tmp.prev.next = null;
			this.front = tmp.prev;
			tmp = null;
		}
		this.nodeCount--;
	}
	void print_queue() {
		node tmp = this.rear;
		while(tmp != null) {
			System.out.print(tmp.data + " ");
			tmp = tmp.next;
		}
		System.out.println('\n');
	}
}

public class mainClass {
	public static void main(String[] args) {
		int arr[] = {5, 10, 15, 20, 2, 4, 6, 8, 20};
		Deque d = new Deque();
		
		for(int i = 0; i < arr.length; i++)
			d.rear_Enqueue(arr[i]);
		d.print_queue();
		
		d.rear_Dequeue();
		d.print_queue();
		d.front_Dequeue();
		d.print_queue();
		
		d.front_Enqueue(90);
		d.print_queue();
	}
}
